module/lifecycle/lifecycle.go

```diff
@@ -49,3 +49,77 @@ func AllDone(components ...module.ReadyDoneAware) <-chan struct{} {
 
 	return done
 }
+
+// LifecycleManager is a support struct for implementing module.ReadyDoneAware
+type LifecycleManager struct {
+	stateTransition   sync.Mutex    // lock for preventing concurrent state transitions
+	started           chan struct{} // used to signal that startup has completed
+	stopped           chan struct{} // used to signal that shutdown has completed
+	startupCommenced  bool          // indicates whether OnStart() has been invoked
+	shutdownCommenced bool          // indicates whether OnStop() has been invoked
+}
+
+func NewLifecycleManager() *LifecycleManager {
+	return &LifecycleManager{
+		stateTransition:   sync.Mutex{},
+		startupCommenced:  false,
+		started:           make(chan struct{}),
+		shutdownCommenced: false,
+		stopped:           make(chan struct{}),
+	}
+}
+
+// OnStart will commence startup of the LifecycleManager. If OnStop has already been called
+// before the first call to OnStart, startup will not be performed. After the first call,
+// subsequent calls to OnStart do nothing.
+func (lm *LifecycleManager) OnStart(startupFns ...func()) {
+	lm.stateTransition.Lock()
+	if lm.shutdownCommenced || lm.startupCommenced {
+		lm.stateTransition.Unlock()
+		return
+	}
+	lm.startupCommenced = true
+	lm.stateTransition.Unlock()
```

Comment:
Thanks! In this particular case, I think it may introduce a bit more messiness than your suggested change, because when we return on line 79 we actually want to return from the enclosing function.

If we wrap the logic in a nested function, the return now only returns from the nested function, so we would probably have to add some if / else logic and return a boolean from the nested function instead. 

I think that may not be necessary here, since the locking code is very short and contained